whois-api/src/main/java/net/ripe/db/whois/api/rest/ClientCertificateExtractor.java

```diff
@@ -0,0 +1,65 @@
+package net.ripe.db.whois.api.rest;
+
+import net.ripe.db.whois.common.io.ByteArrayInput;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServletRequest;
+import java.security.Principal;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+
+import static org.apache.commons.codec.binary.Base64.decodeBase64;
+
+public class ClientCertificateExtractor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClientCertificateExtractor.class);
+
+    private final static String HEADER_SSL_CLIENT_CERT = "SSL_CLIENT_CERT";
+    private final static String HEADER_SSL_CLIENT_VERIFY = "SSL_CLIENT_VERIFY";
+
+    public static Optional<X509Certificate> getClientCertificate(final HttpServletRequest request) {
+        final String sslClientCert = request.getHeader(HEADER_SSL_CLIENT_CERT);
+
+        if (StringUtils.isBlank(sslClientCert)) {
+            return Optional.empty();
+        }
+
+        final String sslClientVerify = request.getHeader(HEADER_SSL_CLIENT_VERIFY);
+        if (!"GENEROUS".equals(sslClientVerify) && !"SUCCESS".equals(sslClientVerify)) {
+            return Optional.empty();
+        }
+
+        return getX509Certificate(sslClientCert);
+    }
+
+    private static Optional<X509Certificate> getX509Certificate(final String certificate) {
+        // TODO what to log as cert identifier (fingerprint, subjectDN?)
+        Principal subject = null;
+        try {
+            final X509Certificate x509 = (X509Certificate) CertificateFactory.getInstance("X.509").generateCertificate(new ByteArrayInput(decodeBase64(certificate)));
+
+            subject = x509.getSubjectDN();
+            x509.checkValidity();
+
+
+            // TODO signature verification (both self signed and ca)
```

Comment:
Remove this?

Reply:
Ok, as discussed we leave the verification to the proxy.